doc/iterations/2024.3/shapes/configurable_collapse.adoc

```diff
@@ -0,0 +1,47 @@
+= Configurable collapse
+
+== Problem
+
+Sirius Web doesn't provide a way for specifiers to set the default collapsible state of a node. It also doesn't provide a way to let specifiers define tools that collapse/expand nodes as part of other operations (e.g. to expand a collapsed node when a child node is created).
+
+Nodes are always created expanded, and the only way to collapse them right now is that the end user does it manually.
+
+== Key Result
+
+- A specifier must be able to define NodeDescriptions that are collapsed/expanded by default.
+- A specifier can define tools that collapse/expand other nodes.
+
+== Solution
+
+A new field _Default Collapsing State Expression_ should be displayed in NodeDescription. The field contains an expression which is evaluated on each diagram render to collapse/expand a node.
+
+This expression would be used to set the Collapsing State if it hasn't been manually set by the end user or updated by a tool.
+
+A new operation _Set Collapsing State_ should be available in tools to expand/collapse nodes. This operation would have a _Graphical Candidates Expression_ field used to retrieve the nodes to expand/collapse, as well as a _CollapsingState_ field containing the value to set (Collapsed or Expanded).
```

Comment:
I agree on the solution you propose, I realize my solution based on an operation only was a bit short sighted, and we should provide a tool that can be used in all the context you mention. I have a question to clarify your suggestion:

> Just add a specific variable (to not pollute existing objects with a new service) and add your service on this.

The variable here is an AQL variable right? If this is the case I assume we want a single variable for a variety of services? This shape only discusses collapse/expand, but if we want to add similar services in the future it should be on the same object/on a nested object, right? Something like

```java
public interface ISiriusWebServices {
    void collapse(List<String> nodeIds);
    void expand(List<String> nodeIds);
    void hide(List<String> nodeIds);
    void pin(List<String> nodeIds)
    // etc
}
```
Some services may not make sense, but I want to clarify that you don't want a new variable that would only contain services to expand/collapse.

Reply:
Yes definitely, we will end up with tons of other features in there like `pin`, `hide`, `createView`, etc. The variable will be usable in AQL but it will simply be a new variable in the `VariableManager`. For example, in `org.eclipse.sirius.components.collaborative.diagrams.DiagramCreationService#doRender` we have a custom variable named `environment` (which is not fully finished) but which would let specifiers figure out in which application their code is executed (Sirius Web vs SysOn for example).

You could add there a new variable `siriusWebServices` or something like that which would then be accessible by everybody (programmatic diagram description, AQL expression, Java services, etc). To make its methods available in AQL expressions, you would have to register its services in the instance of `AQLInterpreter` created for diagrams.

When we convert representation descriptions, we provide the same instance of AQLInterpreter to all representation descriptions (see org.eclipse.sirius.components.view.emf.ViewConverter#convert(...)) while you would need to create an instance per representation description converted and for diagrams you would have to register as services the methods of `ISiriusWebServices`.

This makes me realize that this interface is badly named since it should be `ISiriusWebDiagramServices` or `IDiagramServices` with a variable named `diagramServices` for example. We will end up with form-related services, tree-related ones etc in the near future.

Reply:
updated the shape to integrate the feedback (part of the feedback has been moved to an ADR as suggested)